### PR TITLE
Fix for #523: global planner called much more times than expected

### PR DESCRIPTION
**Pull request by corot**
So now planner is called 1 + max_planning_retries in a first try, and then 1 + max_planning_retries after every recovery behavior. Or up to  planner_patiente seconds after every recovery. I think this is the expected operation.

Seems to work fine now, but move_base.cpp code is quite messy, so probably needs at least one independent tester.

### CODE REVIEW

**Comment by ivan-gavran** [comment]
@corot , would you mind explaining the fix? From the diff, it seems as some combination with locks, but I don't understand it completely.

I tested the behavior, and it is as described: the move base is being called only 1+max_planning_retries for each recovery behavior.

**Comment by corot** [comment]
Hi @gergia, sorry for not being clear. All the issue is better explained in issue #523 and this [ROS answer](http://answers.ros.org/question/242629/global-planner-behavior-upon-planning-failure/).

The change is quite minimal, once whitespaces are ignored: Without the change, note that there's nothing preventing to keep [calling the planner](https://github.com/ros-planning/navigation/pull/524/files#diff-d6afbf7b921c7c246a6ee2454dc05be4L630) even if we are not in PLANNING state. The change makes that planning thread only calls the planner while current state is PLANNING, up to max_planning_retries+1 times or during the time allowed by planner_patiente. If not, it is blocked in the wait_for_wake planner_cond_.wait(lock); (saving CPU)

Hope is clearer now

**Comment by corot** [comment]
Hi all, anything preventing this PR to get merged?

**Comment by corot** [comment]
PR also on kinetic

File: move_base/src/move_base.cpp [hunk 1]
```diff
@@ -619,54 +619,59 @@ namespace move_base {
         wait_for_wake = false;
       }
       ros::Time start_time = ros::Time::now();
-
-      //time to plan! get a copy of the goal and unlock the mutex
-      geometry_msgs::PoseStamped temp_goal = planner_goal_;
       lock.unlock();
```
Comment (mikeferguson) [inline move_base/src/move_base.cpp]: This lock.unlock() call should really move with the temp_goal declaration. We should get the goal WHILE under lock. Of course, that means we need to add an "else{ lock.unlock() }" to the "if (state_==PLANNING)" below, which isn't ideal (it's convoluted) but it is at least thread-safe

File: move_base/src/move_base.cpp [hunk 2]
```diff
@@ -619,54 +619,59 @@ namespace move_base {
...
+            state_ = CLEARING;
+            wait_for_wake = true;
+            publishZeroVelocity();
+            recovery_trigger_ = PLANNING_R;
+          }
+
+          lock.unlock();
         }
-        lock.unlock();
       }
```
Comment (mikeferguson) [inline move_base/src/move_base.cpp]: This is where we would add "else { lock.unlock() }".

I did test this lock change, and it looks all good (also tested the whole feature in general).

**Review by mikeferguson — CHANGES_REQUESTED** [review]
@corot -- apologies for the slow responses on these, we should really keep the same locking structure as before, which requires a pair of simple changes. Would be good to also apply these same changes to #539.

**Comment by mikeferguson** [comment]
Thanks! LGTM -- I'll wait for CI to complete again, and then squash-merge for ease of cherry-picking into K+L.

**Review by mikeferguson — APPROVED** [review]

**Comment by DLu** [comment]
I think this PR breaks the ability to plan with positive `planner_frequency`, since it forces the planner to only run when `state=PLANNING`, but the intended behavior of move_base is to be able to replan continuously, even when in `state=CONTROLLING`. Was this tested with positive `planner_frequency`?

**Comment by corot** [comment]
I think you are right :see_no_evil: 

Will try to fix it asap; sorry for the gaffe

**Comment by corot** [comment]
PR #622
Ok, hope this time all works as intended!  @DLu, can you give a try to the PR? Thanks & thanks for detecting my pitfall!

**Pull request merged**